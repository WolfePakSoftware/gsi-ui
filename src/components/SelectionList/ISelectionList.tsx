import { SemanticICONS} from 'semantic-ui-react';

export interface IItemsList{
    text:string
    icon?:SemanticICONS
    value:string
    active?:boolean,
    iconPosition?: 'left' | 'right'
}

export interface ISelctionListProps{
    elements:IItemsList[]
    onSelect(list:IItemsList[]):void,
    header:{
        text:string
        color?:string        
    },
    selection?:{
        textColor:string,
        background:string
    }
}