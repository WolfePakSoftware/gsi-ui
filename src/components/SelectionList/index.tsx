export { VSelectionList } from './SelectionList';
export type { IItemsList } from './ISelectionList';
export type { EmptyData } from './EmptyData';
