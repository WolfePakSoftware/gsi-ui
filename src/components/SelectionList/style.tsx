import React from 'react';
import styled from 'styled-components';
import { List, SemanticCOLORS } from 'semantic-ui-react';

interface ISelectionProps{
    textColor:SemanticCOLORS,
    background:SemanticCOLORS,
    active:boolean
}

export const ListItem = styled(({ active, textColor, background, ...rest }) => <List.Item {...rest} />)`
  background-color: ${(props: ISelectionProps) => (props.background && props.active) ? `${props.background} !Important` : ''};
  color:${(props: ISelectionProps) => (props.textColor && props.active) ? `${props.textColor} !Important` : ''};
`;