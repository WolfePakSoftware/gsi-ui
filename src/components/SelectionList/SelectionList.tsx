import React, {Component} from 'react';
import { SemanticICONS, Label, Segment, List, Icon } from 'semantic-ui-react';
import {IItemsList,ISelctionListProps} from './ISelectionList';
import {ListItem} from './style'



interface ISelctionListState{
    listSelected:IItemsList[]
}

export class SelectionList extends Component<ISelctionListProps, ISelctionListState>{
    constructor(props:ISelctionListProps){
        super(props);
    }

    public state:ISelctionListState = {
        listSelected:this.props.elements.filter(e=>e.active)
    }
    isActive = (elements:IItemsList)=>{
          return  !!this.state.listSelected.find(e=>e.value === elements.value)
    }
    onItemClick = (active:boolean, element:IItemsList)=>{
        const {listSelected} = this.state;
        let new_list:IItemsList[];

        active
        ?
        new_list = listSelected.filter(e=>e.value!==element.value)
        :
        new_list = listSelected.concat(element);

        this.setState({...listSelected, listSelected:new_list})
        this.props.onSelect(new_list);
    }
    render(){
        const {header, elements, selection} = this.props;
        
        return(
            <Segment basic>
                <Label attached="top" style={{backgroundColor:header.color || '#767676'}}>{header.text}</Label>
                <List selection verticalAlign="middle">
                {
                    elements.map(element=>{
                        const active = this.isActive(element);
                        return(
                            <ListItem active={active} key={element.value} {...selection}>
                                <List.Item
                                    value={element.value}
                                    active={active}
                                    onClick={() => {
                                    this.onItemClick(active, element);
                                    }}>                                    
                                    <List.Content>
                                        {
                                            !!element.icon &&
                                            <Icon name={element.icon} style={{float:(element.iconPosition !=='right')?'left':'inherit'}}/>
                                        }
                                        {element.text}
                                    </List.Content>
                                </List.Item>
                            </ListItem>
                    )})
                }
                </List>
            </Segment>
        );
    }
}