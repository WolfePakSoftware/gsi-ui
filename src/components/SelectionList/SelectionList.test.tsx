import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import {SelectionList} from './SelectionList';
import {IItemsList} from './ISelectionList'

it('renders without crashing Selection List', () => {
    const div = document.createElement('div');
    const array_elements: IItemsList[]= [{text:'first', value:'firstv', icon:'folder'}, {text:'second', value:'secondv', active:true}];
    
    ReactDOM.render(<SelectionList
        elements={array_elements}
        header={{text:'Header', color:'#fbbd08'}}
        onSelect={(list:any)=>console.log(list)}
        selection={{background:'#00b5ad', textColor:'#fbbd08'}}
        />, div);
    ReactDOM.unmountComponentAtNode(div);
});

it('render correctly SelectionList component one segment', () => {
    const array_elements: IItemsList[] = [
        {text:'first', value:'firstv', icon:'folder'}
    ];
    
    const TextInputComponent = renderer
      .create(<SelectionList
        elements={array_elements}
        header={{text:'Header', color:'#fbbd08'}}
        onSelect={(list:any)=>console.log(list)}
        selection={{background:'#00b5ad', textColor:'#fbbd08'}}
        />)
      .toJSON();
    expect(TextInputComponent).toMatchSnapshot();
  });

  it('render correctly SelectionList component two segments', () => {
    const array_elements: IItemsList[] = [{text:'first', value:'firstv', icon:'folder'}, {text:'second', value:'secondv', active:true}];
    
    const TextInputComponent = renderer
      .create(<SelectionList
        elements={array_elements}
        header={{text:'Header', color:'#fbbd08'}}
        onSelect={(list:any)=>console.log(list)}
        selection={{background:'#00b5ad', textColor:'#fbbd08'}}
        />)
      .toJSON();
    expect(TextInputComponent).toMatchSnapshot();
  });