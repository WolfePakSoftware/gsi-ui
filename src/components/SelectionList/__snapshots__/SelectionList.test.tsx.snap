// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly SelectionList component one segment 1`] = `
<div
  className="ui basic segment"
>
  <div
    className="ui top attached label"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#fbbd08",
      }
    }
  >
    Header
  </div>
  <div
    className="ui selection middle aligned list"
    role="list"
  >
    <div
      className="item sc-bdVaJa jeCsUV"
      onClick={[Function]}
      role="listitem"
    >
      <div
        className="item"
        data-value="firstv"
        onClick={[Function]}
        role="listitem"
      >
        <div
          className="content"
        >
          <i
            aria-hidden="true"
            className="folder icon"
            onClick={[Function]}
            style={
              Object {
                "float": "left",
              }
            }
          />
          first
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`render correctly SelectionList component two segments 1`] = `
<div
  className="ui basic segment"
>
  <div
    className="ui top attached label"
    onClick={[Function]}
    style={
      Object {
        "backgroundColor": "#fbbd08",
      }
    }
  >
    Header
  </div>
  <div
    className="ui selection middle aligned list"
    role="list"
  >
    <div
      className="item sc-bdVaJa jeCsUV"
      onClick={[Function]}
      role="listitem"
    >
      <div
        className="item"
        data-value="firstv"
        onClick={[Function]}
        role="listitem"
      >
        <div
          className="content"
        >
          <i
            aria-hidden="true"
            className="folder icon"
            onClick={[Function]}
            style={
              Object {
                "float": "left",
              }
            }
          />
          first
        </div>
      </div>
    </div>
    <div
      className="item sc-bdVaJa ffBesj"
      onClick={[Function]}
      role="listitem"
    >
      <div
        className="active item"
        data-value="secondv"
        onClick={[Function]}
        role="listitem"
      >
        <div
          className="content"
        >
          second
        </div>
      </div>
    </div>
  </div>
</div>
`;
